File: cpp-linux/zugferd_facturx_xrechnung/extract_invoice/extract_invoice.cpp
// extract_invoice -- C++ mirror of
// examples\Vb6\zugferd_facturx_xrechnung\extract_invoice\extract_invoice.bas
// Creates FacturX and XRechnung invoices (attaching factur-x.xml from a memory
// buffer via AttachFileEx) and then verifies the embedded e-invoice can be found
// and extracted again.
#include "apputil.h"
#include <vector>
#include <fstream>

static const unsigned long STD_OUTPUT = 0xFFFFFFF5; // STD_OUTPUT_HANDLE (-11)

static void SetColor(unsigned short attr) {
    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT), attr);
}
enum { COL_RED = 12, COL_GREEN = 10, COL_YELLOW = 14, COL_WHITE = 15 };

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType) {
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

static bool GetFileBuffer(const char* fileName, std::vector<char>& buf) {
    std::ifstream f(fileName, std::ios::binary);
    if (!f) return false;
    f.seekg(0, std::ios::end);
    std::streamoff sz = f.tellg();
    f.seekg(0, std::ios::beg);
    if (sz <= 0) return false;
    buf.resize((size_t)sz);
    f.read(buf.data(), sz);
    return true;
}

static bool HaveEInvoice(PPDF pdf, const char* inFileName) {
    bool ok = false;
    TPDFVersionInfo info; memset(&info, 0, sizeof(info));
    info.StructSize = sizeof(info);

    pdfCreateNewPDFA(pdf, "");
    pdfSetImportFlags(pdf, ifDocInfo | ifEmbeddedFiles);
    pdfSetImportFlags2(pdf, if2UseProxy);

    if (pdfOpenImportFileA(pdf, inFileName, ptOpen, "") < 0) { pdfFreePDF(pdf); return false; }
    pdfImportCatalogObjects(pdf);

    if (pdfGetPDFVersionEx(pdf, &info) == 0) { pdfFreePDF(pdf); return false; }
    if (info.PDFAVersion != 3 || info.FXDocName == 0) { pdfFreePDF(pdf); return false; }

    int ef = pdfFindEmbeddedFileA(pdf, info.FXDocName);
    if (ef < 0) {
        SetColor(COL_RED);
        printf("Invoice %s not found!\n", info.FXDocName);
        pdfFreePDF(pdf);
        return false;
    }
    if (ef != 0) {
        SetColor(COL_YELLOW);
        printf("Warning: The invoice should be the first file attachment. This might cause unnecessary problems.\n");
    }
    // ef + 1: pdfFindEmbeddedFile returns a 0-BASED INDEX while every other
    // embedded-file call takes the 1-BASED handle. That asymmetry is deliberate
    // and documented -- docs\Manual\V17_ExportAPIs\E03_EmbeddedFiles.md says it
    // in as many words ("Unlike every other embedded-file call, this is not the
    // 1-based handle -- add 1 before passing the result to pdfGetEmbeddedFile")
    // and an earlier session already "fixed" the engine to return I+1 and then
    // REVERTED it after reading that page.
    //
    // This example passed the raw index straight through, so it always asked for
    // the wrong file and printed "XML Invoice not found!" even once the
    // attachment really was there. The warning just above already treats the
    // value as 0-based (0 == the first attachment) -- only the +1 was missed.
    TPDFFileSpec fs; memset(&fs, 0, sizeof(fs));
    if (pdfGetEmbeddedFile(pdf, ef + 1, &fs, 1) != 0)
        ok = (fs.BufSize > 0);
    pdfFreePDF(pdf);
    return ok;
}

static bool CreateInvoice(PPDF pdf, bool FacturX, const char* InvoiceName, const char* OutFile) {
    bool result = false;
    pdfCreateNewPDFA(pdf, "");
    pdfSetDocInfoA(pdf, diProducer, "");

    if (pdfOpenImportFileA(pdf, LUMAS_REPO_ROOT "/sample_invoice.pdf", ptOpen, "") < 0) { pdfFreePDF(pdf); return false; }
    pdfImportPDFFile(pdf, 1, 1, 1);

    std::vector<char> buffer;
    int ef;
    if (GetFileBuffer(LUMAS_REPO_ROOT "/factur-x.xml", buffer))
        ef = pdfAttachFileExA(pdf, buffer.data(), (UI32)buffer.size(), InvoiceName, "EN 19631 compliant invoice", 0);
    else
        ef = pdfAttachFileExA(pdf, 0, 0, InvoiceName, "EN 19631 compliant invoice", 0);

    if (FacturX) {
        pdfSetPDFVersion(pdf, pvFacturX_Comfort);
        pdfAssociateEmbFile(pdf, adCatalog, -1, arAlternative, ef);
    } else {
        pdfSetPDFVersion(pdf, pvFacturX_XRechnung);
        pdfAssociateEmbFile(pdf, adCatalog, -1, arSource, ef);
    }

    if (pdfHaveOpenDoc(pdf) != 0) {
        if (pdfOpenOutputFileA(pdf, OutFile) != 0)
            result = (pdfCloseFile(pdf) != 0);
    }
    pdfFreePDF(pdf);
    return result;
}

int main() {
    ChdirToExe();
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, PDFError);

    const char* outFile = "out.pdf";

    if (!CreateInvoice(pdf, true, "factur-x.xml", outFile) || !HaveEInvoice(pdf, outFile) ||
        !CreateInvoice(pdf, false, "xrechnung.xml", outFile) || !HaveEInvoice(pdf, outFile)) {
        SetColor(COL_RED);
        printf("XML Invoice not found!\n");
    } else {
        SetColor(COL_GREEN);
        printf("All tests passed!\n");
    }
    SetColor(COL_WHITE);

    pdfDeletePDF(pdf);
    return 0;
}

File: python/zugferd_facturx_xrechnung/extract_invoice/extract_invoice.py
import sys, os, ctypes
import os, sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

# extract_invoice -- Python (ctypes) port of the VB6 mirror.
# Creates FacturX and XRechnung invoices (attaching factur-x.xml from a memory
# buffer via AttachFileEx) and then verifies the embedded e-invoice can be
# found and extracted again.

HERE = os.path.dirname(os.path.abspath(__file__))
TF = os.path.join(HERE, "..", "..", "..", "test_files")

# console colour helpers (kernel32)
_k32 = ctypes.windll.kernel32
STD_OUTPUT_HANDLE = -11

clRed = 0x0000FF
clGreen = 0x008000
clYellow = 0x00FFFF
clWhite = 0xFFFFFF


def _set_color(color):
    h = _k32.GetStdHandle(STD_OUTPUT_HANDLE)
    _k32.SetConsoleTextAttribute(h, 7)
    attr = {clRed: 12, clGreen: 10, clYellow: 14, clWhite: 15}.get(color)
    if attr is not None:
        _k32.SetConsoleTextAttribute(h, attr)


def _err(data, code, msg, typ):
    if msg:
        print(msg.decode('latin-1', 'replace'))
    return 0


_cb = L.TErrorProc(_err)  # keep global ref


def _ptr_to_str(p):
    if not p:
        return ""
    return ctypes.cast(p, ctypes.c_char_p).value.decode('latin-1', 'replace')


def have_einvoice(pdf, in_file_name):
    info = L.TPDFVersionInfo()
    info.StructSize = ctypes.sizeof(info)
    fs = L.TPDFFileSpec()

    result = False
    L.pdfCreateNewPDFA(pdf, b"")
    L.pdfSetImportFlags(pdf, L.ifDocInfo | L.ifEmbeddedFiles)
    L.pdfSetImportFlags2(pdf, L.if2UseProxy)

    if L.pdfOpenImportFileA(pdf, in_file_name.encode('latin-1'), L.ptOpen, b"") < 0:
        L.pdfFreePDF(pdf)
        return False

    L.pdfImportCatalogObjects(pdf)

    if L.pdfGetPDFVersionEx(pdf, ctypes.byref(info)) == 0:
        L.pdfFreePDF(pdf)
        return False

    if (info.PDFAVersion != 3) or (not info.FXDocName):
        L.pdfFreePDF(pdf)
        return False

    doc_name = _ptr_to_str(info.FXDocName)
    ef = L.pdfFindEmbeddedFileA(pdf, doc_name.encode('latin-1'))
    if ef < 0:
        _set_color(clRed)
        print("Invoice " + doc_name + " not found!")
        L.pdfFreePDF(pdf)
        return False
    if ef != 0:
        _set_color(clYellow)
        print("Warning: The invoice should be the first file attachment. This might cause unnecessary problems.")
    if L.pdfGetEmbeddedFile(pdf, ef, ctypes.byref(fs), 1) != 0:
        result = fs.BufSize > 0
    L.pdfFreePDF(pdf)
    return result


def create_invoice(pdf, facturx, invoice_name, out_file):
    result = False
    L.pdfCreateNewPDFA(pdf, b"")
    L.pdfSetDocInfoA(pdf, L.diProducer, b"")

    if L.pdfOpenImportFileA(pdf, os.path.join(TF, "test_invoice.pdf").encode('latin-1'), L.ptOpen, b"") < 0:
        L.pdfFreePDF(pdf)
        return False

    L.pdfImportPDFFile(pdf, 1, 1.0, 1.0)

    # Override the attachment name (XRechnung requires xrechnung.xml).
    try:
        with open(os.path.join(TF, "factur-x.xml"), "rb") as fh:
            buffer = fh.read()
    except OSError:
        buffer = b""

    if buffer:
        ef = L.pdfAttachFileExA(pdf, buffer, len(buffer), invoice_name.encode('latin-1'),
                                b"EN 19631 compliant invoice", 0)
    else:
        ef = L.pdfAttachFileExA(pdf, 0, 0, invoice_name.encode('latin-1'),
                                b"EN 19631 compliant invoice", 0)

    if facturx:
        L.pdfSetPDFVersion(pdf, L.pvFacturX_Comfort)
        L.pdfAssociateEmbFile(pdf, L.adCatalog, -1, L.arAlternative, ef)
    else:
        L.pdfSetPDFVersion(pdf, L.pvFacturX_XRechnung)
        L.pdfAssociateEmbFile(pdf, L.adCatalog, -1, L.arSource, ef)

    if L.pdfHaveOpenDoc(pdf) != 0:
        if L.pdfOpenOutputFileA(pdf, out_file.encode('latin-1')) != 0:
            result = L.pdfCloseFile(pdf) != 0
    L.pdfFreePDF(pdf)
    return result


def main():
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, _cb)

    outFile = os.path.join(HERE, "out.pdf.")

    if (not create_invoice(pdf, True, "factur-x.xml", outFile)) or (not have_einvoice(pdf, outFile)) \
       or (not create_invoice(pdf, False, "xrechnung.xml", outFile)) or (not have_einvoice(pdf, outFile)):
        _set_color(clRed)
        print("XML Invoice not found!")
    else:
        _set_color(clGreen)
        print("All tests passed!")

    _set_color(clWhite)
    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()
